tests/server.c

```diff
@@ -0,0 +1,359 @@
+#include <io.h> /* write() */
+#include <fcntl.h> /* O_BINARY */
+#else
+#include <sys/socket.h>
+#include <netinet/in.h>
+#include <netdb.h>
+#include <arpa/inet.h>
+#include <unistd.h>
+#include <fcntl.h>
+#endif /* _WIN32 */
+
+#include <sys/types.h>
+#include <stdio.h>
+#include <string.h>
+#include <stdlib.h>
+#include <errno.h>
+#include <errno.h>
+#include <pthread.h>
+#include <time.h>
+
+#ifdef _WIN32
+#define sleep(s) Sleep(1000 * (s))
+#define read(s, buf, n) recv(s, buf, n, 0)
+#define close(s) closesocket(s)
+#define bzero(buf, n) memset(buf, '\0', n)
+
+/* Hacks for 'errno' stuff
+ */
+#undef EAGAIN
+#define EAGAIN WSAEWOULDBLOCK
+#undef EWOULDBLOCK
+#define EWOULDBLOCK WSAEWOULDBLOCK
+#undef errno
+#define errno WSAGetLastError()
+#define perror(str) fprintf(stderr, str ": %d.\n", WSAGetLastError())
+#define strerror(e) ws_strerror(e)
+#ifndef STDOUT_FILENO
+#define STDOUT_FILENO 1 /* MinGW has this */
+#endif /* !STDOUT_FILENO */
+#endif /* _WIN32 */
+
+/* crustls.h is autogenerated in the Makefile using cbindgen. */
+#include "crustls.h"
+#include "common.h"
+
+enum crustls_demo_result
+read_file(const char *filename, char *buf, size_t buflen, size_t *n)
+{
+  FILE *f = fopen(filename, "r");
+  if(f == NULL) {
+    fprintf(stderr, "%s\n", strerror(errno));
+    return CRUSTLS_DEMO_ERROR;
+  }
+  *n = fread(buf, 1, buflen, f);
+  if(!feof(f)) {
+    fprintf(stderr, "%s\n", strerror(errno));
+    return CRUSTLS_DEMO_ERROR;
+  }
+  return CRUSTLS_DEMO_OK;
+}
+
+typedef enum exchange_state {
+  READING_REQUEST,
+  SENT_RESPONSE
+} exchange_state;
+
+/*
+ * Do one read from the socket, and process all resulting bytes into the
+ * rustls_connection, then copy all plaintext bytes from the session to stdout.
+ * Returns:
+ *  - CRUSTLS_DEMO_OK for success
+ *  - CRUSTLS_DEMO_AGAIN if we got an EAGAIN or EWOULDBLOCK reading from the
+ *    socket
+ *  - CRUSTLS_DEMO_EOF if we got EOF
+ *  - CRUSTLS_DEMO_ERROR for other errors.
+ */
+enum crustls_demo_result
+do_read(struct conndata_t *conn, struct rustls_connection *rconn)
+{
+  int err = 1;
+  int result = 1;
+  size_t n = 0;
+
+  err = rustls_connection_read_tls(rconn, read_cb, conn, &n);
+  if(err == EAGAIN || err == EWOULDBLOCK) {
+    fprintf(stderr,
+            "reading from socket: EAGAIN or EWOULDBLOCK: %s\n",
+            strerror(errno));
+    return CRUSTLS_DEMO_AGAIN;
+  }
+  else if(err != 0) {
+    fprintf(stderr, "reading from socket: errno %d\n", err);
+    return CRUSTLS_DEMO_ERROR;
+  }
+
+  if (n == 0) {
+    return CRUSTLS_DEMO_EOF;
+  }
+  fprintf(stderr, "read %ld bytes from socket\n", n);
+
+  result = rustls_connection_process_new_packets(rconn);
+  if(result != RUSTLS_RESULT_OK) {
+    print_error("in process_new_packets", result);
+    return CRUSTLS_DEMO_ERROR;
+  }
+
+  result = copy_plaintext_to_buffer(conn);
+  if(result != CRUSTLS_DEMO_CLOSE_NOTIFY) {
+    fprintf(stderr, "do_read returning %d\n", result);
+    return result;
+  }
+
+  char buf[2048];
```

Comment:
Interesting that you insist on declaring and initializing `err` and `result` at the top of the function like we're back in 1989 but allocate 2k on the stack ~20 lines into the function!